### ISSUE TITLE
Add Enumerable.AggregateRight

### ISSUE THREAD

**Issue by leandromoh** [dotnet/corefx#16458]
# Background

LINQ offers a [Aggregate](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Aggregate.cs) (`foldl`) method, but it does not provide its rigth-associative version  (`foldr`).

![foldr](https://cloud.githubusercontent.com/assets/11452028/23327019/ed4cefbe-fae3-11e6-92ea-63efb9ac0f75.png)

# Proposal

I would like to submit a PR for this, but i dont know exactly which files I must change. There is something beyond to create AggregateRight and tests?

**Comment by karelz** [comment]
What is the value of 'AggregateRight'? How often is it used by applications?

Doesn't Aggregate silently expect the function to be symmetric? (or whatever the right mathematical term is)
Documentation of Aggregate doesn't say anything regarding Left/Right, so I expect it is up to its implementation and it is implementation detail.

Simple workaround is to call '.Reverse().Aggregate(...)'

**Comment by leandromoh** [comment]
> What is the value of 'AggregateRight'? How often is it used by applications?

In functional program the `foldl (aggregate) `and `foldr (aggregateRight)` are so used as `filter (where)` and `map (select)` . Being so power full, that we can define `where `and `select `in its terms.

> Doesn't Aggregate silently expect the function to be symmetric? (or whatever the right mathematical term is)

No, the "direction" of function matters. in a list of number, for example
if function is right-associative, you must use `foldr `to the list be evaluated as `(1 f (2 f (3 f (4 f 5))))`, 
if it is left-associative you must use `foldl `for it be evaluated as `((((1 f 2) f 3) f 4) f 5)`. 
if the function is "symmetric", you can use one of both. see [wikipedia](https://en.wikipedia.org/wiki/Fold_(higher-order_function)#Folds_on_lists), of course  it explains much better than me.

> Documentation of Aggregate doesn't say anything regarding Left/Right, so I expect it is up to its implementation and it is implementation detail.

I saw, its strange no regarding. 

> Simple workaround is to call '.Reverse().Aggregate(...)'

If you program in Haskell, Lisp, Scala, Standard ML, Smalltalk (languages with `foldr`, among others, it isn't equivalent.

If you use `Enumerable.Range(1, 13).Reverse().Aggregate((a, b) => string.Format("({0}/{1})", a, b))`  the result is:
`((((((((((((13/12)/11)/10)/9)/8)/7)/6)/5)/4)/3)/2)/1)
`

Whereas if you also program in one of above languages you will expect:
`(1/(2/(3/(4/(5/(6/(7/(8/(9/(10/(11/(12/13))))))))))))
`

Although the parenteses order is right, the order of the arguments is not, what will give you a complete unexpected result; 

If you are passing a lambda as argument of course you can change the arguments order within it (e.g.,`(a, b) => string.Format("({0}/{1})", b, a)` in the original example). but for functions already declared (e.g., class members, framework's functions, etc) you will need to create a wrapper (e.g., `(b, a) => f(a, b)`). It works, but of course `source.Reverse().Aggregate((a, b) => f(b,a))` is much more ugly than simply call `source.AggregateRight(f)`.

In short, make sense to me create the `AggregateRight` function the way I mentioned (I need it few days ago in a real case). Other point, but lesser important, is that if the collection implements `IList<T>`, is much more performatic to make a reverse `for`than call `Reverse()`, what can be checked within `AggregateRight`.

**Comment by karelz** [comment]
I don't think Linq tries to be full functional language. It provides helpers for common tasks in functional-style syntax and approach.

I think `AggregateRight` would be confusing to many .NET developers who don't care about these subtle nuances. I also believe it would have very marginal usage. Those who truly care can create their own extension methods.
As such I recommend to reject the API proposal, but I will let area experts to make the final call.

**Comment by svick** [comment]
> I would like to submit a PR for this, but i dont know exactly which files I must change. There is something beyond to create AggregateRight and tests?

It's not as simple as that. Any changes to the API have to go though [the API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md).

> I [needed] it few days ago in a real case

Maybe sharing that use case could help make your argument more convincing?

> if the collection implements `IList<T>`, [it] is much more [performant] to make a reverse `for` than call `Reverse()`, what can be checked within `AggregateRight`

Not necessarily. `ImmutableList<T>` implements `IList<T>`, but its indexer is O(log n), so using `for` for reverse iteration is slower for that type.
